tests/utils/test_task_group.py

```diff
@@ -576,3 +579,372 @@ def test_task_without_dag():
+        t_start = PythonOperator(task_id='task_start', python_callable=task_start, dag=dag)
+
+        with TaskGroup("section_1", tooltip="section_1") as section_1:
+            sec_2 = section_2(t_start.output)
+            task_s1 = DummyOperator(task_id="task_1")
+            task_s2 = BashOperator(task_id="task_2", bash_command='echo 1')
+            task_s3 = DummyOperator(task_id="task_3")
+
+            sec_2.set_downstream(task_s1)
+            task_s1 >> [task_s2, task_s3]
+
+        t_end = PythonOperator(task_id='task_end', python_callable=task_end, dag=dag)
+        t_start >> section_1 >> t_end
+
+    node_ids = {
+        'id': None,
+        'children': [
+            {
+                'id': 'section_1',
+                'children': [
+                    {
+                        'id': 'section_1.section_2',
+                        'children': [
+                            {'id': 'section_1.section_2.task_1'},
+                            {'id': 'section_1.section_2.task_2'},
+                            {'id': 'section_1.section_2.task_3'},
+                            {'id': 'section_1.section_2.downstream_join_id'},
+                        ],
+                    },
+                    {'id': 'section_1.task_1'},
+                    {'id': 'section_1.task_2'},
+                    {'id': 'section_1.task_3'},
+                    {'id': 'section_1.upstream_join_id'},
+                    {'id': 'section_1.downstream_join_id'},
+                ],
+            },
+            {'id': 'task_end'},
+            {'id': 'task_start'},
+        ],
+    }
+
+    assert extract_node_id(task_group_to_dict(dag.task_group)) == node_ids
+
+
+def test_duplicate_task_group_id():
+    """ Testing automatic suffix assignment for duplicate group_id"""
+
+    from airflow.operators.python import task
+
+    @task(task_id='start_task')
+    def task_start():
+        """Dummy Task which is First Task of Dag """
+        print('[Task_start]')
+
+    @task(task_id='end_task')
+    def task_end():
+        """Dummy Task which is Last Task of Dag"""
+        print('[Task_End]')
+
+    # Creating Tasks
+    @task(task_id='task')
+    def task_1():
+        """ Dummy Task1"""
+        print('[Task1]')
+
+    @task(task_id='task')
+    def task_2():
+        """ Dummy Task2"""
+        print('[Task2]')
+
+    @task(task_id='task1')
+    def task_3():
+        """ Dummy Task3"""
+        print('[Task3]')
+
+    @task_group_decorator(group_id='task_group1')
+    def task_group1():
+        task_start()
+        task_1()
+        task_2()
+
+    @task_group_decorator(group_id='task_group1')
+    def task_group2():
+        task_3()
+
+    @task_group_decorator(group_id='task_group1')
+    def task_group3():
+        task_end()
+
+    execution_date = pendulum.parse("20201109")
+    with DAG(dag_id="example_duplicate_task_group_id", start_date=execution_date, tags=["example"]) as dag:
+        task_group1()
+        task_group2()
```

Comment:
I would have expected this to fail: this is three _separate_ task groups, all with the same name.

Calling the same TG more than once is when I would expect the `__1` suffix to be added, but having _actual_ duplicate TG ids for different groups I think should be an error case.

Reply:
```
    node_ids = {
        'id': None,
        'children': [
            {
                'id': 'task_group1',
                'children': [
                    {'id': 'task_group1.start_task'},
                    {'id': 'task_group1.task'},
                    {'id': 'task_group1.task__1'},
                ],
            },
            {'id': 'task_group1__1', 'children': [{'id': 'task_group1__1.task1'}]},
            {'id': 'task_group1__2', 'children': [{'id': 'task_group1__2.end_task'}]},
        ],
    }
```

I think this is the correct functionality. With taskgroup decorators people might be importing these from external sources and might therefore not have control over the group_id of that TaskGroup. Automatically adding a suffix can prevent failed DAGs when using these externally created resources.

Reply:
Hmmm, given there's no way to override the name of a TG, true.